Core/include/Acts/Utilities/TypeDispatcher.hpp

```diff
@@ -0,0 +1,193 @@
+// This file is part of the ACTS project.
+//
+// Copyright (C) 2016 CERN for the benefit of the ACTS project
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+#pragma once
+
+#include <format>
+#include <functional>
+#include <type_traits>
+#include <typeindex>
+#include <typeinfo>
+#include <unordered_map>
+#include <vector>
+
+#include <boost/core/demangle.hpp>
+
+namespace Acts {
+
+/// Template class for type-based function dispatch
+///
+/// This class allows registering function pointers associated with specific
+/// derived types of a base class. When invoked with a base class reference,
+/// it will look up and call the appropriate registered function.
+///
+/// @tparam base_t The base class type that will be the first parameter
+/// @tparam signature_t The function signature (e.g., return_t(Args...))
+template <typename base_t, typename signature_t>
+class TypeDispatcher;
+
+template <typename base_t, typename return_t, typename... args_t>
+class TypeDispatcher<base_t, return_t(args_t...)> {
+ public:
+  // Type aliases for frequently used template parameters
+  using base_type = base_t;
+  using return_type = return_t;
+  using self_type = TypeDispatcher<base_type, return_type(args_t...)>;
+
+  using function_signature = return_t(const base_t&, args_t...);
+  using function_type = std::function<function_signature>;
+
+  /// Default constructor
+  TypeDispatcher() = default;
+
+  /// Constructor that registers multiple function pointers with auto-detected
+  /// types
+  /// @param funcs Function pointers - derived types are auto-detected from their first parameters
+  template <typename... derived_types_t>
+  explicit TypeDispatcher(return_t (*... funcs)(const derived_types_t&,
+                                                args_t...))
+    requires(std::is_base_of_v<base_t, derived_types_t> && ...)
+  {
+    (registerFunction(funcs),
+     ...);  // Fold expression to register each function
+  }
+
+  /// Register a free function with explicit derived type
+  /// @tparam derived_t The derived type to associate the function with
+  /// @param func The function pointer
+  template <typename derived_t>
+    requires std::is_base_of_v<base_t, derived_t>
+  self_type& registerFunction(return_t (*func)(const derived_t&, args_t...)) {
+    std::type_index typeIdx(typeid(derived_t));
```

Comment:
Doe I read this correctly:
- with this mechanism, it is only the derived type that determines the index, because only exactly one function per derive type can be registered, the fact that - in your example - the functions are named the same is convention, but not enforced, right?

Reply:
It's possible to have more than one function that "works" for a given derived type, but that's an error case. I try to detect this in the registration (if derived is default-constructible) or at call time: if more than one `dynamic_cast` is successful, that's an exception. The idea is that only one will "engage".